File: service/node_service.py
import datetime

from utils.ed25519 import *
from model.node import Node


class NodeService:
    def __init__(self):
        pass

    @staticmethod
    def create(identifier: str, display_name: str, description: str, open: bool, creator: str) -> dict:
        if Node.select().where(Node.identifier == identifier).exists():
            raise Exception(f"Node {identifier} already exists")

        signing_private_key, signing_public_key = generate_ed25519_keys()

        node = Node(
            identifier=identifier,
            display_name=display_name,
            description=description,
            open=open,
            signing_private_key=private_key_to_string(signing_private_key),
            signing_public_key=public_key_to_string(signing_public_key),
            creator=creator
        )

        node.save()

        return {
            'id': node.get_id(),
            'identifier': identifier
        }

    @staticmethod
    def fetch(page=0, size=50) -> list[dict]:
        nodes = Node.select().paginate(page, size)
        result = []
        for node in nodes:
            result.append(NodeService.to_dict(node))
        return result

    @staticmethod
    def fetch_open(page=0, size=50) -> list[dict]:
        nodes = Node.select().where(Node.open == True).paginate(page, size)
        result = []
        for node in nodes:
            result.append(NodeService.to_dict(node))
        return result

    @staticmethod
    def get(identifier: str):
        node = Node.select().where(Node.identifier == identifier).get_or_none()

        if not node:
            raise Exception(f"Node {identifier} not found")

        return NodeService.to_dict(node)

    @staticmethod
    def update(identifier: str, display_name: str, description: str) -> dict:
        count = Node.update(
            identifier=identifier,
            display_name=display_name,
            description=description,
            updated_at=datetime.datetime.now(tz=datetime.timezone.utc)
        ).where(Node.identifier == identifier).execute()

        if count == 0:
            raise Exception(f"Node {identifier} not found")

        return {
            "identifier": identifier,
        }

    @staticmethod
    def update_open(identifier: str, open: bool) -> dict:
        count = Node.update(
            open=open,
            updated_at=datetime.datetime.now(tz=datetime.timezone.utc)
        ).where(Node.identifier == identifier).execute()

        if count == 0:
            raise Exception(f"Node {identifier} not found")

        return {
            "identifier": identifier,
        }

    @staticmethod
    def delete(identifier: str) -> dict:
        count = Node.delete().where(Node.identifier == identifier).execute()

        if count == 0:
            raise Exception(f"Node {identifier} not found")

        return {
            "identifier": identifier
        }

    @staticmethod
    def reset_signing_keys(identifier: str) -> dict:
        signing_private_key, signing_public_key = generate_ed25519_keys()

        count = Node.update(
            signing_private_key=private_key_to_string(signing_private_key),
            signing_public_key=public_key_to_string(signing_public_key),
            updated_at=datetime.datetime.now(tz=datetime.timezone.utc)
        ).where(Node.identifier == identifier).execute()

        if count == 0:
            raise Exception(f"Node {identifier} not found")

        return {
            "identifier": identifier,
            "signing_public_key": public_key_to_string(signing_public_key),
        }

    @staticmethod
    def to_dict(node: Node) -> dict:
        return {
            'id': node.get_id(),
            'identifier': node.identifier,
            'display_name': node.display_name,
            'description': node.description,
            'open': node.open,
            'signing_public_key': node.signing_public_key,
            'creator': node.creator,
            'created_at': node.created_at,
            'updated_at': node.updated_at,
        }

File: internal/app/vertex/node/node_service.go
package node

import (
	"fmt"
	"github.com/evernetproto/evernet/internal/pkg/keys"
	"gorm.io/gorm"
)

type Service struct {
	db *gorm.DB
}

func NewService(db *gorm.DB) *Service {
	return &Service{db: db}
}

func (s *Service) Create(request *CreationRequest, creator string) (*Node, error) {
	var count int64

	err := s.db.Model(&Node{}).Where(Node{Identifier: request.Identifier}).Count(&count).Error

	if err != nil {
		return nil, err
	}

	if count != 0 {
		return nil, fmt.Errorf("node %s already exists", request.Identifier)
	}

	signingPublicKey, signingPrivateKey, err := keys.GenerateKeyPair()

	if err != nil {
		return nil, err
	}

	signingPrivateKeyString := keys.PrivateKeyToString(signingPrivateKey)
	signingPublicKeyString := keys.PublicKeyToString(signingPublicKey)

	node := &Node{
		Identifier:        request.Identifier,
		DisplayName:       request.DisplayName,
		Description:       request.Description,
		SigningPublicKey:  signingPublicKeyString,
		SigningPrivateKey: signingPrivateKeyString,
		Open:              request.Open,
		Creator:           creator,
	}

	err = s.db.Create(node).Error

	if err != nil {
		return nil, err
	}

	return node, nil
}

func (s *Service) List(page int, size int) ([]*Node, error) {
	nodes := make([]*Node, 0)

	err := s.db.Limit(size).Offset(page * size).Find(&nodes).Error

	if err != nil {
		return nil, err
	}

	return nodes, nil
}

func (s *Service) ListOpen(page int, size int) ([]*Node, error) {
	nodes := make([]*Node, 0)

	err := s.db.Limit(size).Offset(page * size).Where(Node{Open: true}).Find(&nodes).Error

	if err != nil {
		return nil, err
	}

	return nodes, nil
}

func (s *Service) Get(identifier string) (*Node, error) {
	node := &Node{}

	err := s.db.Where(Node{Identifier: identifier}).First(node).Error

	if err != nil {
		return nil, err
	}

	return node, nil
}

func (s *Service) Update(identifier string, request *UpdateRequest) (*Node, error) {
	node, err := s.Get(identifier)

	if err != nil {
		return nil, err
	}

	if request.DisplayName != "" {
		node.DisplayName = request.DisplayName
	}

	node.Description = request.Description

	err = s.db.Save(node).Error

	if err != nil {
		return nil, err
	}

	return node, nil
}

func (s *Service) UpdateOpen(identifier string, request *OpenUpdateRequest) (*Node, error) {
	node, err := s.Get(identifier)

	if err != nil {
		return nil, err
	}

	node.Open = request.Open

	err = s.db.Save(node).Error

	if err != nil {
		return nil, err
	}

	return node, nil
}

func (s *Service) Delete(identifier string) (*Node, error) {
	node, err := s.Get(identifier)

	if err != nil {
		return nil, err
	}

	err = s.db.Delete(node).Error

	if err != nil {
		return nil, err
	}

	return node, nil
}

func (s *Service) ResetSigningKeys(identifier string) (*Node, error) {
	node, err := s.Get(identifier)

	if err != nil {
		return nil, err
	}

	signingPublicKey, signingPrivateKey, err := keys.GenerateKeyPair()

	if err != nil {
		return nil, err
	}

	signingPrivateKeyString := keys.PrivateKeyToString(signingPrivateKey)
	signingPublicKeyString := keys.PublicKeyToString(signingPublicKey)

	node.SigningPublicKey = signingPublicKeyString
	node.SigningPrivateKey = signingPrivateKeyString

	err = s.db.Save(node).Error

	if err != nil {
		return nil, err
	}

	return node, err
}
